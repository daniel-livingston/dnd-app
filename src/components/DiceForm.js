import React from "react";
import { connect } from "react-redux";
import Form from "./Form";
import {
	addToHistory,
	rollDice,
	updateDiceNumber,
	updateDiceModifier,
	updateDiceSides,
} from "../actions/dice";
import { DEFAULT_DICE_NUMBERS, DEFAULT_DICE_SIDES } from "../constants";

export class DiceForm extends React.Component {
	onNumberSelect = (e) => {
		const value = parseInt(e.target.value, 10);
		this.props.updateNumber(value);
	};
	onSidesSelect = (e) => {
		const value = parseInt(e.target.value, 10);
		this.props.updateSides(value);
	};
	onModifierChange = (e) => {
		const value = parseInt(e.target.value, 10);
		this.props.updateModifier(value);
	};
	onSubmit = () => {
		document.getElementById("DiceForm-Button").setAttribute("disabled", "disabled");

		const rolls = [];
		let total = 0;
		for (let i = 0; i < this.props.number; i++) {
			const roll = Math.floor(Math.random() * this.props.sides + 1);
			rolls.push(roll);
			total += roll;
		}
		total += this.props.modifier;

		const roll = {
			rolls,
			number: this.props.number,
			sides: this.props.sides,
			modifier: this.props.modifier,
			total,
			time: new Date(),
			name: this.props.name || "Guest",
		};

		this.props.rollDice(roll);
		setTimeout(() => {
			document.getElementById("DiceForm-Button").removeAttribute("disabled");
		}, 100);
	};
	render() {
		return (
			<Form
				className='DiceForm'
				header='Dice Roller'
				items={[
					{
						title: "Number",
						description: "Choose the number of dice to roll.",
						inputType: "select",
						inputValue: this.props.number,
						inputOnChange: this.onNumberSelect,
						inputPlaceholder: "Number",
						selectOptions: DEFAULT_DICE_NUMBERS,
					},
					{
						title: "Sides",
						description: "Choose the number of sides for the dice.",
						inputType: "select",
						inputValue: this.props.sides,
						inputOnChange: this.onSidesSelect,
						inputPlaceholder: "Sides",
						selectOptions: DEFAULT_DICE_SIDES,
					},
					{
						title: "Modifier",
						description: "Choose the modifier for the roll.",
						inputType: "number",
						inputValue: this.props.modifier,
						inputOnChange: this.onModifierChange,
						inputPlaceholder: "Modifier",
					},
				]}
				button={{
					id: "DiceForm-Button",
					onClick: this.onSubmit,
					name: "Roll dice",
				}}
			/>
		);
	}
}

const mapStateToProps = (state) => ({
	number: state.dice.number,
	sides: state.dice.sides,
	modifier: state.dice.modifier,
	last: state.dice.last,
	name: state.user.username,
});

const mapDispatchToProps = (dispatch) => ({
	addToHistory: (roll) => dispatch(addToHistory(roll)),
	rollDice: (roll) => dispatch(rollDice(roll)),
	updateNumber: (number) => dispatch(updateDiceNumber(number)),
	updateSides: (sides) => dispatch(updateDiceSides(sides)),
	updateModifier: (modifier) => dispatch(updateDiceModifier(modifier)),
});

export default connect(mapStateToProps, mapDispatchToProps)(DiceForm);
