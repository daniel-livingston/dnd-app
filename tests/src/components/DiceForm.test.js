import React from "react";
import { Provider } from "react-redux";
import { mount } from "enzyme";
import DiceForm from "../../../src/components/DiceForm";
import store from "../../../src/store/configureStore";

// TODO: Test functions

describe("<DiceForm>", () => {
	test("renders correctly", () => {
		const wrapper = mount(
			<Provider store={store}>
				<DiceForm />
			</Provider>
		);
		expect(wrapper).toMatchSnapshot();
	});
});
