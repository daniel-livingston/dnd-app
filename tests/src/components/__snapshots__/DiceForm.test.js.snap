// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DiceForm> renders correctly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(DiceForm)>
    <DiceForm
      addToHistory={[Function]}
      last={Object {}}
      modifier={0}
      number={1}
      rollDice={[Function]}
      sides={20}
      updateModifier={[Function]}
      updateNumber={[Function]}
      updateSides={[Function]}
    >
      <Form
        button={
          Object {
            "id": "DiceForm-Button",
            "name": "Roll dice",
            "onClick": [Function],
          }
        }
        className="DiceForm"
        header="Dice Roller"
        items={
          Array [
            Object {
              "description": "Choose the number of dice to roll.",
              "inputOnChange": [Function],
              "inputPlaceholder": "Number",
              "inputType": "select",
              "inputValue": 1,
              "selectOptions": Array [
                1,
                2,
                3,
                4,
                5,
                6,
              ],
              "title": "Number",
            },
            Object {
              "description": "Choose the number of sides for the dice.",
              "inputOnChange": [Function],
              "inputPlaceholder": "Sides",
              "inputType": "select",
              "inputValue": 20,
              "selectOptions": Array [
                4,
                6,
                8,
                10,
                12,
                20,
                100,
              ],
              "title": "Sides",
            },
            Object {
              "description": "Choose the modifier for the roll.",
              "inputOnChange": [Function],
              "inputPlaceholder": "Modifier",
              "inputType": "number",
              "inputValue": 0,
              "title": "Modifier",
            },
          ]
        }
      >
        <form
          className="Form"
          onSubmit={[Function]}
        >
          <h2
            className="Form-header"
          >
            Dice Roller
          </h2>
          <FormItem
            description="Choose the number of dice to roll."
            inputOnChange={[Function]}
            inputPlaceholder="Number"
            inputType="select"
            inputValue={1}
            key="Number"
            selectOptions={
              Array [
                1,
                2,
                3,
                4,
                5,
                6,
              ]
            }
            title="Number"
          >
            <div
              className="FormItem"
            >
              <FormItemInfo
                description="Choose the number of dice to roll."
                title="Number"
              >
                <div
                  className="FormItemInfo"
                >
                  <h3
                    className="FormItemInfo-Title"
                  >
                    Number
                  </h3>
                  <p
                    className="FormItemInfo-Description"
                  >
                    Choose the number of dice to roll.
                  </p>
                </div>
              </FormItemInfo>
              <FormItemInput
                onChange={[Function]}
                options={
                  Array [
                    1,
                    2,
                    3,
                    4,
                    5,
                    6,
                  ]
                }
                placeholder="Number"
                type="select"
                value={1}
              >
                <select
                  className="FormItemInput FormItemInput-Select"
                  onChange={[Function]}
                  value={1}
                >
                  <option
                    key="1"
                    value={1}
                  >
                    1
                  </option>
                  <option
                    key="2"
                    value={2}
                  >
                    2
                  </option>
                  <option
                    key="3"
                    value={3}
                  >
                    3
                  </option>
                  <option
                    key="4"
                    value={4}
                  >
                    4
                  </option>
                  <option
                    key="5"
                    value={5}
                  >
                    5
                  </option>
                  <option
                    key="6"
                    value={6}
                  >
                    6
                  </option>
                </select>
              </FormItemInput>
            </div>
          </FormItem>
          <FormItem
            description="Choose the number of sides for the dice."
            inputOnChange={[Function]}
            inputPlaceholder="Sides"
            inputType="select"
            inputValue={20}
            key="Sides"
            selectOptions={
              Array [
                4,
                6,
                8,
                10,
                12,
                20,
                100,
              ]
            }
            title="Sides"
          >
            <div
              className="FormItem"
            >
              <FormItemInfo
                description="Choose the number of sides for the dice."
                title="Sides"
              >
                <div
                  className="FormItemInfo"
                >
                  <h3
                    className="FormItemInfo-Title"
                  >
                    Sides
                  </h3>
                  <p
                    className="FormItemInfo-Description"
                  >
                    Choose the number of sides for the dice.
                  </p>
                </div>
              </FormItemInfo>
              <FormItemInput
                onChange={[Function]}
                options={
                  Array [
                    4,
                    6,
                    8,
                    10,
                    12,
                    20,
                    100,
                  ]
                }
                placeholder="Sides"
                type="select"
                value={20}
              >
                <select
                  className="FormItemInput FormItemInput-Select"
                  onChange={[Function]}
                  value={20}
                >
                  <option
                    key="4"
                    value={4}
                  >
                    4
                  </option>
                  <option
                    key="6"
                    value={6}
                  >
                    6
                  </option>
                  <option
                    key="8"
                    value={8}
                  >
                    8
                  </option>
                  <option
                    key="10"
                    value={10}
                  >
                    10
                  </option>
                  <option
                    key="12"
                    value={12}
                  >
                    12
                  </option>
                  <option
                    key="20"
                    value={20}
                  >
                    20
                  </option>
                  <option
                    key="100"
                    value={100}
                  >
                    100
                  </option>
                </select>
              </FormItemInput>
            </div>
          </FormItem>
          <FormItem
            description="Choose the modifier for the roll."
            inputOnChange={[Function]}
            inputPlaceholder="Modifier"
            inputType="number"
            inputValue={0}
            key="Modifier"
            title="Modifier"
          >
            <div
              className="FormItem"
            >
              <FormItemInfo
                description="Choose the modifier for the roll."
                title="Modifier"
              >
                <div
                  className="FormItemInfo"
                >
                  <h3
                    className="FormItemInfo-Title"
                  >
                    Modifier
                  </h3>
                  <p
                    className="FormItemInfo-Description"
                  >
                    Choose the modifier for the roll.
                  </p>
                </div>
              </FormItemInfo>
              <FormItemInput
                onChange={[Function]}
                placeholder="Modifier"
                type="number"
                value={0}
              >
                <input
                  className="FormItemInput FormItemInput-Number"
                  onChange={[Function]}
                  type="number"
                  value={0}
                />
              </FormItemInput>
            </div>
          </FormItem>
          <FormButton
            id="DiceForm-Button"
            name="Roll dice"
            onClick={[Function]}
          >
            <div
              className="FormButton-Container"
              style={
                Object {
                  "justifyContent": "flex-end",
                }
              }
            >
              <Button
                id="DiceForm-Button"
                name="Roll dice"
                onClick={[Function]}
                width="auto"
              >
                <button
                  className="Button"
                  id="DiceForm-Button"
                  onClick={[Function]}
                  style={
                    Object {
                      "width": "auto",
                    }
                  }
                >
                  Roll dice
                </button>
              </Button>
            </div>
          </FormButton>
        </form>
      </Form>
    </DiceForm>
  </Connect(DiceForm)>
</Provider>
`;
